### PR TITLE
RavenDB-10815 - Nagle support in session

### PR DESCRIPTION
**Pull request by grisha-kotler**
http://issues.hibernatingrhinos.com/issue/RavenDB-10815

### CODE REVIEW

File: Raven.Client.Lightweight/Document/DocumentStore.cs [hunk 1]
```diff
@@ -825,6 +909,174 @@ public Task GetObserveChangesAndEvictItemsFromCacheTask(string database = null)
 
             return changes == null ? new CompletedTask() : changes.ConnectionTask;
         }
+
+        public class NagleData
```
Comment (ppekrol) [inline Raven.Client.Lightweight/Document/DocumentStore.cs]: can this be private or internal?

File: Raven.Client.Lightweight/Document/DocumentStore.cs [hunk 2]
```diff
@@ -825,6 +909,174 @@ public Task GetObserveChangesAndEvictItemsFromCacheTask(string database = null)
...
+        {
+            public InMemoryDocumentSessionOperations.SaveChangesData SaveChangesData { get; set; }
+
+            public TaskCompletionSource<BatchResult[]> TaskCompletionSource = new TaskCompletionSource<BatchResult[]>();
+        }
+
+        private readonly BlockingCollection<NagleData> nagleQueue = new BlockingCollection<NagleData>();
+        private Task nagleWriterTask;
+
+        public Task<BatchResult[]> AddNagleData(InMemoryDocumentSessionOperations.SaveChangesData saveChangesData)
```
Comment (ppekrol) [inline Raven.Client.Lightweight/Document/DocumentStore.cs]: can this be private or internal?

File: Raven.Client.Lightweight/Document/DocumentStore.cs (hunk 1 above)
Comment (grisha-kotler) [reply Raven.Client.Lightweight/Document/DocumentStore.cs]: done

File: Raven.Client.Lightweight/Document/DocumentStore.cs (hunk 2 above)
Comment (grisha-kotler) [reply Raven.Client.Lightweight/Document/DocumentStore.cs]: done

File: Raven.Client.Lightweight/Document/DocumentStore.cs [hunk 3]
```diff
@@ -825,6 +909,174 @@ public Task GetObserveChangesAndEvictItemsFromCacheTask(string database = null)
...
+                batchOptions.ThrowOnTimeoutInWaitForReplicas |= option.ThrowOnTimeoutInWaitForReplicas;
+
+                batchOptions.WaitForIndexes |= option.WaitForIndexes;
+                batchOptions.WaitForIndexesTimeout =
+                    batchOptions.WaitForIndexesTimeout > option.WaitForIndexesTimeout
+                        ? batchOptions.WaitForIndexesTimeout
+                        : option.WaitForIndexesTimeout;
+
+                batchOptions.ThrowOnTimeoutInWaitForIndexes |= option.ThrowOnTimeoutInWaitForIndexes;
+                if (option.WaitForSpecificIndexes != null)
```
Comment (ayende) [inline Raven.Client.Lightweight/Document/DocumentStore.cs:1065]: Here this is wrong. 
If we have a request that wait for specific indexes, and one that waits for all, we need to wait for all

Comment (grisha-kotler) [reply Raven.Client.Lightweight/Document/DocumentStore.cs:1065]: done

**Pull request merged**